### ISSUE TITLE
Improve Swagger Generation for complicated Authenticators (and OAuth/OpenIDConnect)

### ISSUE THREAD

**Issue by airstandley** [plangrid/flask-rebar#125]
Came across another fun Authentication use case that Rebar doesn't currently support.

OpenAPI 2 and 3 both supported AND/OR security requirements. https://swagger.io/docs/specification/authentication/#multiple

The example they give is an API that requires both an API Key and an App ID. 

Writing an Authenticator that does this is simple:

```
from flask import request, g
from flask-rebar import HeaderAPIKeyAuthenticator, errors


class MyAuthenticator(HeaderAPIKeyAuthenticator):
    def __init__(api_key_header, app_id_header, name="myAuth")
        super(MyAuthenticator, self).__init__(api_key_header, name=name)
        self.app_id_header = app_id_header

    def authenticate(self):
        super(MyAuthenticator, self).authenticate()
        if self.app_id_header not in request.headers:
            raise errors.Unauthorized("Missing application id")
        g.authenticated_app_id = request.headers.get(self.app_id_header)


authenticator = MyAuthenticator("X-API-Key", "X-App-ID")
```

The correct swagger spec is also pretty simple

```
"securitySchemes": {
    "apiKey": {
        "type": "apiKey",
        "in": "header",
        "name": "X-API-Key"
    },
    "appID": {
       "type": "apiKey",
        "in": "header",
        "name": "X-App-ID"
    },
}

"security": [
    {"apiKey": [], "appID": []},
]
```

However you're basically stuck writing your own SwaggerGenerator if you want to generate such a spec since our AuthenticatorConverter only supports {name: []} SecurityRequirements and only supports one SecurityDefinition per Authenticator.

Also that {name:[]} SecurityRequirements assumption means that we can't generate valid requirements for OAuth and OpenIDConnect since those should include scope names in the list.

I'm proposing switching to Authenticators registering a AuthenticatorConverter object rather than a single method. An AuthenticatorConverter should have `get_security_schemas` and `get_security_requirements` methods that accept an Authenticator instance, and return a list of Security Scheme Objects and Security Requirement Objects respectively

**Comment by Sytten** [comment]
I don't believe we support AND in our authenticators. But we do support the OR now. What I would really like to add is something like what spring does with the @PreAuthorize (https://docs.spring.io/spring-security/site/docs/3.0.x/reference/el-access.html).
The use case I face most of the time for a path is I want either an admin role OR a user role WITH some permissions (including a check if the ID in the security context match what is in the path).

**Comment by airstandley** [comment]
My rationale for AND is that an Authenticator can be complex enough that the only way to represent the security requirement in swagger is with an AND of two security definitions. See the example above. 
I'm not aware of any way to represent that security requirement in swagger without using AND, please do let me know if I've missed something!

I think extending Authenticators to accept the handler arguments is a great idea, and shouldn't be too hard to add for version 2.0. You should file an issue for that!

**Comment by Sytten** [comment]
I totally agree with you, we absolutely need the AND (we can't do that currently). My point was that we might want more complex auth with multiple AND/OR. I am not sure how to do that properly though (except with composition like how you did it).

**Comment by airstandley** [comment]
Fixed by #130
